<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Spatially-aware quality control (QC) software for both spot-level 
    and artifact-level QC in spot-based spatial transcripomics, such as 10x 
    Visium. These methods calculate local (nearest-neighbors) mean and variance 
    of standard QC metrics (library size, unique genes, and mitochondrial 
    percentage) to identify outliers spot and large technical artifacts. Scales
    linearly with the number of spots and is designed to be used with
    SpatialExperiment objects.">
<title>Spatially-aware quality control for spatial transcriptomics â€¢ SpotSweeper</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Spatially-aware quality control for spatial transcriptomics">
<meta property="og:description" content="Spatially-aware quality control (QC) software for both spot-level 
    and artifact-level QC in spot-based spatial transcripomics, such as 10x 
    Visium. These methods calculate local (nearest-neighbors) mean and variance 
    of standard QC metrics (library size, unique genes, and mitochondrial 
    percentage) to identify outliers spot and large technical artifacts. Scales
    linearly with the number of spots and is designed to be used with
    SpatialExperiment objects.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">SpotSweeper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/getting_started.html">Getting started with 'SpotSweeper'</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MicTott/SpotSweeper/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="spotsweeper">SpotSweeper<a class="anchor" aria-label="anchor" href="#spotsweeper"></a>
</h1></div>
<!-- badges: start -->

<p><code>SpotSweeper</code> is a package developed for spatially-aware quality control (QC) methods for the detection, visualization, and removal of both local outliers and large artifacts in spot-based spatial transcriptomics data, such as 10x Genomics <code>Visium</code>, using standard QC metrics.</p>
<div class="section level2">
<h2 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h2>
<p>Currently, the only way to install <code>SpotSweeper</code> is by downloading the development version which can be installed from <a href="https://github.com/MicTott/SpotSweeper" class="external-link">GitHub</a> using the following:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"devtools"</span>)) <span class="fu">install.packages</span>(<span class="st">"devtools"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"MicTott/SpotSweeper"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="er">}</span></span></code></pre></div>
<p>Once accepted in <a href="http://bioconductor.org/" class="external-link">Bioconductor</a>, <code>SpotSweeper</code> will be installable using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SpotSweeper"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorial-in-progress">Tutorial (in progress)<a class="anchor" aria-label="anchor" href="#tutorial-in-progress"></a>
</h2>
<p>A detailed tutorial is available in the package vignette from Bioconductor. A direct link to the tutorial / package vignette is available here.</p>
</div>
<div class="section level2">
<h2 id="input-data-format">Input data format<a class="anchor" aria-label="anchor" href="#input-data-format"></a>
</h2>
<p>In the examples below, we assume the input data are provided as a <a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a> (SPE) object. The outputs for spot-level outliers and artifacts are stored in the <code>colData</code> of the SPE object.</p>
</div>
<div class="section level2">
<h2 id="spot-level-local-outlier-detection">Spot-level local outlier detection<a class="anchor" aria-label="anchor" href="#spot-level-local-outlier-detection"></a>
</h2>
<p>This is an example workflow showing how to detect and visualize local outliers in 10X Genomics Visium data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MicTott/SpotSweeper" class="external-link">SpotSweeper</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># load  Maynard et al DLPFC daatset</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">STexampleData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/STexampleData/man/STexampleData.html" class="external-link">Visium_humanDLPFC</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?STexampleData and browseVignettes('STexampleData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span></span>
<span><span class="co"># change from gene id to gene names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># show column data before SpotSweeper</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcode_id"   "sample_id"    "in_tissue"    "array_row"    "array_col"   </span></span>
<span><span class="co">#&gt; [6] "ground_truth" "cell_count"</span></span>
<span></span>
<span><span class="co"># drop out-of-tissue spots</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="va">spe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>SpotSweeper can be run on an SPE object with the following code. This outputs the <code>local_outliers</code> in the colData of the SPE object. Selecting <code>data_output=TRUE</code> exports z-transformed QC metrics as well.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Identifying the mitochondrial transcripts in our SpatialExperiment.</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Calculating QC metrics for each spot using scuttle</span></span>
<span><span class="va">spe</span><span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">spe</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode_id"            "sample_id"             "in_tissue"            </span></span>
<span><span class="co">#&gt;  [4] "array_row"             "array_col"             "ground_truth"         </span></span>
<span><span class="co">#&gt;  [7] "cell_count"            "sum"                   "detected"             </span></span>
<span><span class="co">#&gt; [10] "subsets_Mito_sum"      "subsets_Mito_detected" "subsets_Mito_percent" </span></span>
<span><span class="co">#&gt; [13] "total"</span></span>
<span></span>
<span><span class="co"># Identifying local outliers using SpotSweeper</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                         metric<span class="op">=</span><span class="st">"sum"</span>,</span>
<span>                         direction<span class="op">=</span><span class="st">"lower"</span>,</span>
<span>                         log<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                         metric<span class="op">=</span><span class="st">"detected"</span>,</span>
<span>                         direction<span class="op">=</span><span class="st">"lower"</span>,</span>
<span>                         log<span class="op">=</span><span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                         metric<span class="op">=</span><span class="st">"subsets_Mito_percent"</span>,</span>
<span>                         direction<span class="op">=</span><span class="st">"higher"</span>,</span>
<span>                         log<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine all outliers into "local_outliers" column</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">local_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sum_outliers</span><span class="op">)</span> <span class="op">|</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">detected_outliers</span><span class="op">)</span> <span class="op">|</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">subsets_Mito_percent_outliers</span><span class="op">)</span></span></code></pre></div>
<p>We can now visualize <code>local_outliers</code> vs one of the QC metrics, <code>sum_log2</code>, with help from the <code>escheR</code> package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/boyiguo1/escheR" class="external-link">escheR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># library size</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum_log2"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"sum_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># unique genes</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"detected_log2"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"detected_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Unique Genes"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mitochondrial percent</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"subsets_Mito_percent"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"subsets_Mito_percent_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mitochondrial Percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># all local outliers</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum_log2"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"local_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span>, stroke<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"All Local Outliers"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>  plotlist <span class="op">=</span> <span class="va">plot_list</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">2</span>, nrow <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  common.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-local_outlier_plot-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="artifact-detection">Artifact detection<a class="anchor" aria-label="anchor" href="#artifact-detection"></a>
</h2>
<p>Large artifacts can be visualized and detected by calculating the local variance of standard QC metrics, such as mitochondrial ratio.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># load in DLPFC sample with hangnail artifact</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DLPFC_artifact</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">DLPFC_artifact</span></span>
<span></span>
<span><span class="co"># inspect colData before artifact detection</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "sample_id"          "in_tissue"          "array_row"         </span></span>
<span><span class="co">#&gt;  [4] "array_col"          "key"                "sum_umi"           </span></span>
<span><span class="co">#&gt;  [7] "sum_gene"           "expr_chrM"          "expr_chrM_ratio"   </span></span>
<span><span class="co">#&gt; [10] "ManualAnnotation"   "subject"            "region"            </span></span>
<span><span class="co">#&gt; [13] "sex"                "age"                "diagnosis"         </span></span>
<span><span class="co">#&gt; [16] "sample_id_complete" "count"              "sizeFactor"</span></span>
<span></span>
<span><span class="co"># find artifacts using </span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/findArtifacts.html">findArtifacts</a></span><span class="op">(</span><span class="va">spe</span>, </span>
<span>                     mito_percent<span class="op">=</span><span class="st">"expr_chrM_ratio"</span>,</span>
<span>                     mito_sum<span class="op">=</span><span class="st">"expr_chrM"</span>,</span>
<span>                     n_rings<span class="op">=</span><span class="fl">5</span>,</span>
<span>                     name<span class="op">=</span><span class="st">"artifact"</span></span>
<span>                    <span class="op">)</span></span>
<span></span>
<span><span class="co"># check that "artifact" is now in colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "sample_id"            "in_tissue"            "array_row"           </span></span>
<span><span class="co">#&gt;  [4] "array_col"            "key"                  "sum_umi"             </span></span>
<span><span class="co">#&gt;  [7] "sum_gene"             "expr_chrM"            "expr_chrM_ratio"     </span></span>
<span><span class="co">#&gt; [10] "ManualAnnotation"     "subject"              "region"              </span></span>
<span><span class="co">#&gt; [13] "sex"                  "age"                  "diagnosis"           </span></span>
<span><span class="co">#&gt; [16] "sample_id_complete"   "count"                "sizeFactor"          </span></span>
<span><span class="co">#&gt; [19] "expr_chrM_ratio_log2" "expr_chrM_log2"       "coords"              </span></span>
<span><span class="co">#&gt; [22] "k6"                   "k18"                  "k36"                 </span></span>
<span><span class="co">#&gt; [25] "k60"                  "k90"                  "Kmeans"              </span></span>
<span><span class="co">#&gt; [28] "artifact"</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># plotting using escheR</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/make_escheR.html" class="external-link">make_escheR</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_fill.html" class="external-link">add_fill</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"expr_chrM_ratio"</span>, point_size<span class="op">=</span><span class="fl">1.25</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/make_escheR.html" class="external-link">make_escheR</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_fill.html" class="external-link">add_fill</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"expr_chrM"</span>, point_size<span class="op">=</span><span class="fl">1.25</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/make_escheR.html" class="external-link">make_escheR</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_fill.html" class="external-link">add_fill</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"k18"</span>, point_size<span class="op">=</span><span class="fl">1.25</span><span class="op">)</span></span>
<span>  </span>
<span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/make_escheR.html" class="external-link">make_escheR</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_fill.html" class="external-link">add_fill</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"artifact"</span>, point_size<span class="op">=</span><span class="fl">1.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">""</span>,</span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Scale for fill is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for fill, which will replace the existing scale.</span></span>
<span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>  plotlist <span class="op">=</span> <span class="va">plot_list</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">2</span>, nrow <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  common.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-artifact_visualization-1.png" width="100%"></p>
</div>
<div class="section level2">
<h2 id="development-tools">Development tools<a class="anchor" aria-label="anchor" href="#development-tools"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="sc">*</span> Continuous code testing is possible thanks to [GitHub actions](https<span class="sc">:</span><span class="er">//</span>www.tidyverse.org<span class="sc">/</span>blog<span class="sc">/</span><span class="dv">2020</span><span class="sc">/</span><span class="dv">04</span><span class="sc">/</span>usethis<span class="dv">-1-6-0</span><span class="sc">/</span>)  through <span class="st">`</span><span class="at">r BiocStyle::Biocpkg('biocthis')</span><span class="st">`</span>.</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="sc">*</span> The [documentation website](http<span class="sc">:</span><span class="er">//</span>MicTott.github.io<span class="sc">/</span>SpotSweeper) is automatically updated thanks to <span class="st">`</span><span class="at">r BiocStyle::CRANpkg('pkgdown')</span><span class="st">`</span>.</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="sc">*</span> The code is styled automatically thanks to <span class="st">`</span><span class="at">r BiocStyle::CRANpkg('styler')</span><span class="st">`</span>.</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="sc">*</span> The documentation is formatted thanks to <span class="st">`</span><span class="at">r BiocStyle::CRANpkg('devtools')</span><span class="st">`</span> and <span class="st">`</span><span class="at">r BiocStyle::CRANpkg('roxygen2')</span><span class="st">`</span>.</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>This package was developed using <span class="st">`</span><span class="at">r BiocStyle::Biocpkg('biocthis')</span><span class="st">`</span>.</span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MicTott/SpotSweeper/" class="external-link">Browse source code</a></li>
<li><a href="https://support.bioconductor.org/tag/SpotSweeper" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
<li><a href="SUPPORT.html">Getting help</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SpotSweeper</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Michael Totty <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-9292-8556" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://github.com/MicTott/SpotSweeper/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/MicTott/SpotSweeper/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael Totty.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
