---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

# SpotSweeper (under development)

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![check-bioc](https://github.com/MicTott/SpotSweeper/actions/workflows/check-bioc.yml/badge.svg)](https://github.com/MicTott/SpotSweeper/actions/workflows/check-bioc.yml)
<!-- badges: end -->

`SpotSweeper` is a BioConductor package for spatially-aware quality control (QC) methods for the detection, visualization, and removal of local outliers in spot-based spatial transcriptomics data, such as 10x Genomics `Visium`, using standard QC metrics. 

## Installation instructions (in progress)

Get the latest stable `R` release from [CRAN](http://cran.r-project.org/). Then install `SpotSweeper` from [Bioconductor](http://bioconductor.org/) using the following code:

```{r 'install', eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("SpotSweeper")
```

And the development version from [GitHub](https://github.com/MicTott/SpotSweeper) with:

```{r 'install_dev', eval = FALSE}
BiocManager::install("MicTott/SpotSweeper")
```

## Tutorial (in progress)

A detailed tutorial is available in the package vignette from Bioconductor. A direct link to the tutorial / package vignette is available here.

## Input data format
In the examples below, we assume the input data are provided as a SpatialExperiment (SPE) object. The outputs for spot-level outliers are stored in the colData of the SPE object.

(FUTURE TODO)
Alternatively, the inputs can also be provided as a numeric matrix of normalized and transformed counts (e.g. log-transformed normalized counts, also known as logcounts) and a numeric matrix of spatial coordinates.

## Spot-level local outlier detection

This is an exmaple workflow showing how to detect and visualize local outliers in 10X Genomics Visium data. 

```{r example_spe, eval = requireNamespace('SpotSweeper')}
library(SpotSweeper)
suppressPackageStartupMessages({
  library(SpatialExperiment)
})


# load  Maynard et al DLPFC daatset
spe <- STexampleData::Visium_humanDLPFC()

# change from gene id to gene names
rownames(spe) <- rowData(spe)$gene_name

# show column data before SpotSweepR
colnames(colData(spe))

# drop out-of-tissue spots
spe <- spe[, spe$in_tissue == 1]
spe <- spe[, !is.na(spe$ground_truth)]
```

SpotSweeper can be run on an SPE object with the following code. This outputs the `local_outliers` in the colData of the SPE object. Selecting `data_output=TRUE` exports z-transformed QC metrics as well.

```{r example_localOutliers, eval = requireNamespace('SpotSweeper')}

# Identifying the mitochondrial transcripts in our SpatialExperiment.
is.mito <- rownames(spe)[grepl("^MT-", rownames(spe))]

# Calculating QC metrics for each spot using scuttle
spe<- scuttle::addPerCellQCMetrics(spe, subsets=list(Mito=is.mito))
colnames(colData(spe))

# Identifying local outliers suing SpotSweepR
features <- c('sum' ,'detected', "subsets_Mito_percent")
spe<- localOutliers(spe, 
                    features=features,
                    n_neighbors=36, 
                    data_output=TRUE,
                    method="multivariate"
                    )

# show column data before SpotSweepR
colnames(colData(spe))

```

Below we can visualize `local_outliers` vs one of the QC metrics, `sum_log2`, using the `escheR` package.

```{r local_outlier_plot}
library(escheR)

# plotting using escheR
make_escheR(spe) |> 
  add_fill(var = "sum_log2") |>
  add_ground(var = "local_outliers", stroke = 1) +
  scale_color_manual(
    name = "", # turn off legend name for ground_truth
    values = c(
      "TRUE" = "red",
      "FALSE" = "transparent")
  ) +
  scale_fill_gradient(low ="white",high =  "darkgreen")
```

## Region-level artifact detection (in progress)

Regional artifacts, such as "smears" and tissue tears, can be visualized and detected by calculating the local variance of mitochondrial percent or ratio. 

TODO: add example with regional artifact. 

```{r example_localVariance, eval = requireNamespace('SpotSweeper')}

features = c("subsets_Mito_percent")

spe <- localVariance(spe, 
                     features=features,
                     n_neighbors=18, 
                     name="local_mito_variance_k18"
                     )

# show column data after calculating local variance
colnames(colData(spe))

```

```{r local_variance_plot}
library(ggpubr)

# plotting using escheR
p1 <- make_escheR(spe) |> 
  add_fill(var = "subsets_Mito_percent", point_size=1)

p2 <- make_escheR(spe) |> 
  add_fill(var = "local_mito_variance_k18", point_size=1)

plot_list <- list(p1, p2)
ggarrange(
  plotlist = plot_list,
  ncol = 2, nrow = 1,
  common.legend = FALSE)
```

## Citation

Below is the citation output from using `citation('SpotSweeper')` in R. Please
run this yourself to check for any updates on how to cite __SpotSweeper__.

```{r 'citation', eval = requireNamespace('SpotSweeper')}
print(citation("SpotSweeper"), bibtex = TRUE)
```

Please note that the `SpotSweeper` was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.

## Code of Conduct

Please note that the `SpotSweeper` project is released with a [Contributor Code of Conduct](http://bioconductor.org/about/code-of-conduct/). By contributing to this project, you agree to abide by its terms.

## Development tools

* Continuous code testing is possible thanks to [GitHub actions](https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/)  through `r BiocStyle::CRANpkg('usethis')`, `r BiocStyle::CRANpkg('remotes')`, and `r BiocStyle::CRANpkg('rcmdcheck')` customized to use [Bioconductor's docker containers](https://www.bioconductor.org/help/docker/) and `r BiocStyle::Biocpkg('BiocCheck')`.
* Code coverage assessment is possible thanks to [codecov](https://codecov.io/gh) and `r BiocStyle::CRANpkg('covr')`.
* The [documentation website](http://MicTott.github.io/SpotSweeper) is automatically updated thanks to `r BiocStyle::CRANpkg('pkgdown')`.
* The code is styled automatically thanks to `r BiocStyle::CRANpkg('styler')`.
* The documentation is formatted thanks to `r BiocStyle::CRANpkg('devtools')` and `r BiocStyle::CRANpkg('roxygen2')`.

For more details, check the `dev` directory.

This package was developed using `r BiocStyle::Biocpkg('biocthis')`.


