---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

# SpotSweeper

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/MicTott/SpotSweeper/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/MicTott/SpotSweeper/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

`SpotSweeper` is a package developed for spatially-aware quality control (QC) methods for the detection, visualization, and removal of both local outliers and large artifacts in spot-based spatial transcriptomics data, such as 10x Genomics `Visium`, using standard QC metrics. 


## Installation instructions

Currently, the only way to install `SpotSweeper` is by downloading the development version which can be installed from [GitHub](https://github.com/MicTott/SpotSweeper) using the following: 

```{r 'install_dev', eval = FALSE}
if (!require("devtools")) install.packages("devtools")
remotes::install_github("MicTott/SpotSweeper")
}
```

Once accepted in [Bioconductor](http://bioconductor.org/), `SpotSweeper` will be installable using:

```{r 'install', eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("SpotSweeper")
```

## Tutorial (in progress)

A detailed tutorial is available in the package vignette from Bioconductor. A direct link to the tutorial / package vignette is available here.

## Input data format
In the examples below, we assume the input data are provided as a [SpatialExperiment](https://github.com/drighelli/SpatialExperiment) (SPE) object. The outputs for spot-level outliers and artifacts are stored in the `colData` of the SPE object.


## Spot-level local outlier detection

This is an example workflow showing how to detect and visualize local outliers in 10X Genomics Visium data. 

```{r example_spe, eval = requireNamespace('SpotSweeper')}
library(SpotSweeper)


# load  Maynard et al DLPFC daatset
spe <- STexampleData::Visium_humanDLPFC()

# change from gene id to gene names
rownames(spe) <- rowData(spe)$gene_name

# show column data before SpotSweeper
colnames(colData(spe))

# drop out-of-tissue spots
spe <- spe[, spe$in_tissue == 1]
spe <- spe[, !is.na(spe$ground_truth)]
```

SpotSweeper can be run on an SPE object with the following code. This outputs the `local_outliers` in the colData of the SPE object. Selecting `data_output=TRUE` exports z-transformed QC metrics as well.

```{r example_localOutliers, eval = requireNamespace('SpotSweeper')}

# Identifying the mitochondrial transcripts in our SpatialExperiment.
is.mito <- rownames(spe)[grepl("^MT-", rownames(spe))]

# Calculating QC metrics for each spot using scuttle
spe<- scuttle::addPerCellQCMetrics(spe, subsets=list(Mito=is.mito))
colnames(colData(spe))

# Identifying local outliers using SpotSweeper
features <- c('sum' ,'detected', "subsets_Mito_percent")
spe<- localOutliers(spe, 
                    features=features,
                    n_neighbors=18, 
                    data_output=TRUE,
                    method="multivariate"
                    )

# show column data after SpotSweeper
colnames(colData(spe))

```

We can now visualize `local_outliers` vs one of the QC metrics, `sum_log2`, with help from the `escheR` package.

```{r local_outlier_plot}
library(escheR)
library(ggpubr)

# plotting using escheR
p1 <- make_escheR(spe) |> 
  add_fill(var = "sum_log2", point_size=1.25) +
  scale_fill_gradient(low ="white",high =  "darkgreen")

p2 <- make_escheR(spe) |> 
  add_fill(var = "sum_log2", point_size=1.25) |>
  add_ground(var = "local_outliers", stroke = 1) +
  scale_color_manual(
    name = "", # turn off legend name for ground_truth
    values = c(
      "TRUE" = "red",
      "FALSE" = "transparent")
  ) +
  scale_fill_gradient(low ="white",high =  "darkgreen")

plot_list <- list(p1, p2)
ggarrange(
  plotlist = plot_list,
  ncol = 2, nrow = 1,
  common.legend = FALSE
)
```

## Artifact detection

Large artifacts can be visualized and detected by calculating the local variance of standard QC metrics, such as mitochondrial ratio.

```{r example_artifactRemoval, eval = requireNamespace('SpotSweeper')}

# load in DLPFC sample with hangnail artifact
data(DLPFC_artifact)
spe <- DLPFC_artifact

# inspect colData before artifact detection
colnames(colData(spe))

# find artifacts using 
spe <- findArtifacts(spe, 
                     mito_percent="expr_chrM_ratio",
                     mito_sum="expr_chrM",
                     n_rings=5,
                     name="artifact"
                    )

# check that "artifact" is now in colData
colnames(colData(spe))

```

```{r artifact_visualization}

# plotting using escheR
p1 <- make_escheR(spe) |> 
  add_fill(var = "expr_chrM_ratio", point_size=1.25)

p2 <- make_escheR(spe) |> 
  add_fill(var = "expr_chrM", point_size=1.25)

p3 <- make_escheR(spe) |> 
  add_fill(var = "k18", point_size=1.25)
  

p4 <- make_escheR(spe) |> 
  add_fill(var = "artifact", point_size=1.25) +
  scale_fill_manual(
    name = "",
    values = c(
      "TRUE" = "red",
      "FALSE" = "grey")
  )

plot_list <- list(p1, p2, p3, p4)
ggarrange(
  plotlist = plot_list,
  ncol = 2, nrow = 2,
  common.legend = FALSE
)
```



## Citation

Please look out for our preprint describing `SpotSweeper` on bioRxiv. Until then, please cite the package as follows:

```{r citation, eval = FALSE}
citation("SpotSweeper")
```


## Development tools

```{r dev_tools, eval = FALSE}

* Continuous code testing is possible thanks to [GitHub actions](https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/)  through `r BiocStyle::Biocpkg('biocthis')`.
* The [documentation website](http://MicTott.github.io/SpotSweeper) is automatically updated thanks to `r BiocStyle::CRANpkg('pkgdown')`.
* The code is styled automatically thanks to `r BiocStyle::CRANpkg('styler')`.
* The documentation is formatted thanks to `r BiocStyle::CRANpkg('devtools')` and `r BiocStyle::CRANpkg('roxygen2')`.

This package was developed using `r BiocStyle::Biocpkg('biocthis')`.
```
