---
title: "Getting started with 'SpotSweeper'"
author: 
  - name: Michael Totty
    affiliation: &id1 "Johns Hopkins Bloomberg School of Public Health, Baltimore, MD, USA"
  - name: Boyi Guo
    affiliation: *id1
  - name: Stephanie Hicks
    affiliation: *id1
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Getting Started with 'SpotSweeper'}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  message = FALSE,
  comment = "#>"
  #fig.width = 8,
  #fig.height = 8,
  #fig.retina = 2
)
```

## Input data format
In the examples below, we assume the input data are provided as a [SpatialExperiment](https://github.com/drighelli/SpatialExperiment) (SPE) object. The outputs for spot-level outliers and artifacts are stored in the `colData` of the SPE object.


## Spot-level local outlier detection

This is an example workflow showing how to detect and visualize local outliers in 10X Genomics Visium data. 

```{r example_spe, eval = requireNamespace('SpotSweeper')}
library(SpotSweeper)


# load  Maynard et al DLPFC daatset
spe <- STexampleData::Visium_humanDLPFC()

# change from gene id to gene names
rownames(spe) <- rowData(spe)$gene_name

# show column data before SpotSweeper
colnames(colData(spe))

# drop out-of-tissue spots
spe <- spe[, spe$in_tissue == 1]
spe <- spe[, !is.na(spe$ground_truth)]
```

SpotSweeper can be run on an SPE object with the following code. This outputs the `local_outliers` in the colData of the SPE object. Selecting `data_output=TRUE` exports z-transformed QC metrics as well.

```{r example_localOutliers, eval = requireNamespace('SpotSweeper')}

# Identifying the mitochondrial transcripts in our SpatialExperiment.
is.mito <- rownames(spe)[grepl("^MT-", rownames(spe))]

# Calculating QC metrics for each spot using scuttle
spe<- scuttle::addPerCellQCMetrics(spe, subsets=list(Mito=is.mito))
colnames(colData(spe))

# Identifying local outliers using SpotSweeper
spe <- localOutliers(spe,
                         metric="sum",
                         direction="lower",
                         log=TRUE
)

spe <- localOutliers(spe,
                         metric="detected",
                         direction="lower",
                         log=TRUE
)

spe <- localOutliers(spe,
                         metric="subsets_Mito_percent",
                         direction="higher",
                         log=FALSE
)

# combine all outliers into "local_outliers" column
spe$local_outliers <- as.logical(spe$sum_outliers) | 
  as.logical(spe$detected_outliers) | 
  as.logical(spe$subsets_Mito_percent_outliers)

```

We can now visualize `local_outliers` vs one of the QC metrics, `sum_log2`, with help from the `escheR` package.

```{r local_outlier_plot}
library(escheR)
library(ggpubr)

# library size
p1 <- plotOutliers(spe, metric="sum_log2", 
             outliers="sum_outliers", point_size=1.1) +
  ggtitle("Library Size")

# unique genes
p2 <- plotOutliers(spe, metric="detected_log2", 
             outliers="detected_outliers", point_size=1.1) +
  ggtitle("Unique Genes")

# mitochondrial percent
p3 <- plotOutliers(spe, metric="subsets_Mito_percent", 
             outliers="subsets_Mito_percent_outliers", point_size=1.1) +
  ggtitle("Mitochondrial Percent")

# all local outliers
p4 <- plotOutliers(spe, metric="sum_log2", 
             outliers="local_outliers", point_size=1.1, stroke=0.75) +
  ggtitle("All Local Outliers")

# plot
plot_list <- list(p1, p2, p3, p4)
ggarrange(
  plotlist = plot_list,
  ncol = 2, nrow = 2,
  common.legend = FALSE
)

```



# Session information

```{r}
utils::sessionInfo()
```
