<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SpotSweeper">
<title>Getting started with 'SpotSweeper' • SpotSweeper</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with 'SpotSweeper'">
<meta property="og:description" content="SpotSweeper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SpotSweeper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting_started.html">Getting started with 'SpotSweeper'</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MicTott/SpotSweeper/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with 'SpotSweeper'</h1>
                        <h4 data-toc-skip class="author">Michael
Totty</h4>
            <address class="author_afil">
      Johns Hopkins Bloomberg School of Public Health, Baltimore, MD,
USA<br><h4 data-toc-skip class="author">Boyi
Guo</h4>
            <address class="author_afil">
      Johns Hopkins Bloomberg School of Public Health, Baltimore, MD,
USA<br><h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            <address class="author_afil">
      Johns Hopkins Bloomberg School of Public Health, Baltimore, MD,
USA<br><h4 data-toc-skip class="date">2024-03-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/MicTott/SpotSweeper/blob/HEAD/vignettes/getting_started.Rmd" class="external-link"><code>vignettes/getting_started.Rmd</code></a></small>
      <div class="d-none name"><code>getting_started.Rmd</code></div>
    </address>
</address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>SpotSweeper</code> is an R package for spatial transcriptomics
data quality control (QC). It provides functions for detecting and
visualizing spot-level local outliers and artifacts using
spatially-aware methods. The package is designed to work with <a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a>
objects, and is compatible with data from 10X Genomics Visium and other
spatial transcriptomics platforms.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Currently, the only way to install <code>SpotSweeper</code> is by
downloading the development version which can be installed from <a href="https://github.com/MicTott/SpotSweeper" class="external-link">GitHub</a> using the
following:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">"devtools"</span>)) <span class="fu">install.packages</span>(<span class="st">"devtools"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"MicTott/SpotSweeper"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="er">}</span></span></code></pre></div>
<p>Once accepted in <a href="http://bioconductor.org/" class="external-link">Bioconductor</a>,
<code>SpotSweeper</code> will be installable using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SpotSweeper"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="spot-level-local-outlier-detection">Spot-level local outlier detection<a class="anchor" aria-label="anchor" href="#spot-level-local-outlier-detection"></a>
</h2>
<div class="section level3">
<h3 id="loading-example-data">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data"></a>
</h3>
<p>Here we’ll walk you through the standard workflow for using
‘SpotSweeper’ to detect and visualize local outliers in spatial
transcriptomics data. We’ll use the <code>Visium_humanDLPFC</code>
dataset from the <code>STexampleData</code> package, which is a
<code>SpatialExperiment</code> object.</p>
<p>Because local outliers will be saved in the <code>colData</code> of
the <code>SpatialExperiment</code> object, we’ll first view the
<code>colData</code> and drop out-of-tissue spots before calculating
quality control (QC) metrics and running <code>SpotSweeper</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MicTott/SpotSweeper" class="external-link">SpotSweeper</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load  Maynard et al DLPFC daatset</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">STexampleData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/STexampleData/man/STexampleData.html" class="external-link">Visium_humanDLPFC</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show column data before SpotSweeper</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcode_id"   "sample_id"    "in_tissue"    "array_row"    "array_col"   </span></span>
<span><span class="co">#&gt; [6] "ground_truth" "cell_count"</span></span>
<span></span>
<span><span class="co"># drop out-of-tissue spots</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="va">spe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="calculating-qc-metrics-using-scuttle">Calculating QC metrics using <code>scuttle</code><a class="anchor" aria-label="anchor" href="#calculating-qc-metrics-using-scuttle"></a>
</h3>
<p>We’ll use the <code>scuttle</code> package to calculate QC metrics.
To do this, we’ll need to first change the <code>rownames</code> from
gene id to gene names. We’ll then get the mitochondrial transcripts and
calculate QC metrics for each spot using
<code><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">scuttle::addPerCellQCMetrics</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change from gene id to gene names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># identifying the mitochondrial transcripts</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># calculating QC metrics for each spot using scuttle</span></span>
<span><span class="va">spe</span><span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">spe</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode_id"            "sample_id"             "in_tissue"            </span></span>
<span><span class="co">#&gt;  [4] "array_row"             "array_col"             "ground_truth"         </span></span>
<span><span class="co">#&gt;  [7] "cell_count"            "sum"                   "detected"             </span></span>
<span><span class="co">#&gt; [10] "subsets_Mito_sum"      "subsets_Mito_detected" "subsets_Mito_percent" </span></span>
<span><span class="co">#&gt; [13] "total"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="identifying-local-outliers-using-spotsweeper">Identifying local outliers using <code>SpotSweeper</code><a class="anchor" aria-label="anchor" href="#identifying-local-outliers-using-spotsweeper"></a>
</h3>
<p>We can now use <code>SpotSweeper</code> to identify local outliers in
the spatial transcriptomics data. We’ll use the
<code>localOutliers</code> function to detect local outliers based on
the unique detected genes, total library size, and percent of the total
reads that are mitochondrial. These methods assume a normal
distribution, so we’ll use the log-transformed sum of the counts and the
log-transformed number of detected genes. For mitochondrial percent,
we’ll use the raw mitochondrial percentage.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library size</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum"</span>,direction<span class="op">=</span><span class="st">"lower"</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># unique genes</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"detected"</span>, direction<span class="op">=</span><span class="st">"lower"</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mitochondrial percent</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"subsets_Mito_percent"</span>, direction<span class="op">=</span><span class="st">"higher"</span>, log<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The <code>localOutlier</code> function automatically outputs the
results to the <code>colData</code> with the naming convention
<code>X_outliers</code>, where <code>X</code> is the name of the input
<code>colData</code>. We can then combine all outliers into a single
column called <code>local_outliers</code> in the <code>colData</code> of
the <code>SpatialExperiment</code> object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine all outliers into "local_outliers" column</span></span>
<span><span class="va">spe</span><span class="op">$</span><span class="va">local_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">sum_outliers</span><span class="op">)</span> <span class="op">|</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">detected_outliers</span><span class="op">)</span> <span class="op">|</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">subsets_Mito_percent_outliers</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-local-outliers">Visualizing local outliers<a class="anchor" aria-label="anchor" href="#visualizing-local-outliers"></a>
</h3>
<p>We can visualize the local outliers using the
<code>plotOutliers</code> function. This function creates a scatter plot
of the specified metric and highlights the local outliers in red using
the <code>escheR</code> package. Here, we’ll visualize local outliers of
library size, unique genes, mitochondrial percent, and finally, all
local outliers. We’ll then arrange these plots in a grid using
<code>ggpubr::arrange</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/boyiguo1/escheR" class="external-link">escheR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># library size</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum_log2"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"sum_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># unique genes</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"detected_log2"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"detected_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Unique Genes"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mitochondrial percent</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"subsets_Mito_percent"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"subsets_Mito_percent_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mitochondrial Percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># all local outliers</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum_log2"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"local_outliers"</span>, point_size<span class="op">=</span><span class="fl">1.1</span>, stroke<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"All Local Outliers"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>  plotlist <span class="op">=</span> <span class="va">plot_list</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">2</span>, nrow <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  common.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="getting_started_files/figure-html/local_outlier_plot-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="removing-technical-artifacts-using-spotsweeper">Removing technical artifacts using <code>SpotSweeper</code><a class="anchor" aria-label="anchor" href="#removing-technical-artifacts-using-spotsweeper"></a>
</h2>
<div class="section level3">
<h3 id="loading-example-data-1">Loading example data<a class="anchor" aria-label="anchor" href="#loading-example-data-1"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load in DLPFC sample with hangnail artifact</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">DLPFC_artifact</span><span class="op">)</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">DLPFC_artifact</span></span>
<span></span>
<span><span class="co"># inspect colData before artifact detection</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "sample_id"          "in_tissue"          "array_row"         </span></span>
<span><span class="co">#&gt;  [4] "array_col"          "key"                "sum_umi"           </span></span>
<span><span class="co">#&gt;  [7] "sum_gene"           "expr_chrM"          "expr_chrM_ratio"   </span></span>
<span><span class="co">#&gt; [10] "ManualAnnotation"   "subject"            "region"            </span></span>
<span><span class="co">#&gt; [13] "sex"                "age"                "diagnosis"         </span></span>
<span><span class="co">#&gt; [16] "sample_id_complete" "count"              "sizeFactor"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-technical-artifacts">Visualizing technical artifacts<a class="anchor" aria-label="anchor" href="#visualizing-technical-artifacts"></a>
</h3>
<p>Technical artifacts can commonly be visualized by standard QC
metrics, including library size, unique genes, or mitochondrial
percentage. We can first visualize the technical artifacts using the
<code>plotOutliers</code> function. This function plots the Visium spots
with the specified QC metric.We’ll then again arrange these plots using
<code>ggpubr::arrange</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># library size</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum_umi"</span>, </span>
<span>             outliers<span class="op">=</span><span class="cn">NULL</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># unique genes</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"sum_gene"</span>, </span>
<span>             outliers<span class="op">=</span><span class="cn">NULL</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Unique Genes"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># mitochondrial percent</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"expr_chrM_ratio"</span>, </span>
<span>             outliers<span class="op">=</span><span class="cn">NULL</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mitochondrial Percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>  plotlist <span class="op">=</span> <span class="va">plot_list</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">3</span>, nrow <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  common.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="getting_started_files/figure-html/artifact_QC_plots-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="identifying-artifacts-using-spotsweeper">Identifying artifacts using <code>SpotSweeper</code><a class="anchor" aria-label="anchor" href="#identifying-artifacts-using-spotsweeper"></a>
</h3>
<p>We can then use the <code>findArtifacts</code> function to identify
artifacts in the spatial transcriptomics (data. This function identifies
technical artifacts based on the first principle component of the local
variance of the specified QC metric (<code>mito_percent</code>) at
numerous neighorhood sizes (<code>n_rings=5</code>). Currently,
<code>kmeans</code> clustering is used to cluster the technical artifact
vs high-quality Visium spots. Similar to <code>localOutliers</code>, the
<code>findArtifacts</code> function then outputs the results to the
<code>colData</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find artifacts using SpotSweeper</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findArtifacts.html">findArtifacts</a></span><span class="op">(</span><span class="va">spe</span>, </span>
<span>                     mito_percent<span class="op">=</span><span class="st">"expr_chrM_ratio"</span>,</span>
<span>                     mito_sum<span class="op">=</span><span class="st">"expr_chrM"</span>,</span>
<span>                     n_rings<span class="op">=</span><span class="fl">5</span>,</span>
<span>                     name<span class="op">=</span><span class="st">"artifact"</span></span>
<span>                    <span class="op">)</span></span>
<span></span>
<span><span class="co"># check that "artifact" is now in colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "sample_id"            "in_tissue"            "array_row"           </span></span>
<span><span class="co">#&gt;  [4] "array_col"            "key"                  "sum_umi"             </span></span>
<span><span class="co">#&gt;  [7] "sum_gene"             "expr_chrM"            "expr_chrM_ratio"     </span></span>
<span><span class="co">#&gt; [10] "ManualAnnotation"     "subject"              "region"              </span></span>
<span><span class="co">#&gt; [13] "sex"                  "age"                  "diagnosis"           </span></span>
<span><span class="co">#&gt; [16] "sample_id_complete"   "count"                "sizeFactor"          </span></span>
<span><span class="co">#&gt; [19] "expr_chrM_ratio_log2" "expr_chrM_log2"       "coords"              </span></span>
<span><span class="co">#&gt; [22] "k6"                   "k18"                  "k36"                 </span></span>
<span><span class="co">#&gt; [25] "k60"                  "k90"                  "Kmeans"              </span></span>
<span><span class="co">#&gt; [28] "artifact"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-artifacts">Visualizing artifacts<a class="anchor" aria-label="anchor" href="#visualizing-artifacts"></a>
</h3>
<p>We can visualize the artifacts using the <code>escheR</code> package.
Here, we’ll visualize the artifacts using the <code>make_escheR</code>
function and arrange these plots using <code>ggpubr::arrange</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotOutliers.html">plotOutliers</a></span><span class="op">(</span><span class="va">spe</span>, metric<span class="op">=</span><span class="st">"expr_chrM_ratio"</span>, </span>
<span>             outliers<span class="op">=</span><span class="st">"artifact"</span>, point_size<span class="op">=</span><span class="fl">1.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Hangnail artifact"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getting_started_files/figure-html/artifact_visualization-1.png" width="700">
# Session information</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggpubr_0.6.0                escheR_1.2.0               </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.5.0               STexampleData_1.10.1       </span></span>
<span><span class="co">#&gt;  [5] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.54.1        GenomeInfoDb_1.38.8        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.36.0              S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.14.0       matrixStats_1.2.0          </span></span>
<span><span class="co">#&gt; [15] ExperimentHub_2.10.0        AnnotationHub_3.10.0       </span></span>
<span><span class="co">#&gt; [17] BiocFileCache_2.10.1        dbplyr_2.5.0               </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.48.1         SpotSweeper_0.99.1         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.2                     bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.3                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;   [5] compiler_4.3.3                RSQLite_2.3.5                </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.24.0     png_0.1-8                    </span></span>
<span><span class="co">#&gt;   [9] systemfonts_1.0.6             vctrs_0.6.5                  </span></span>
<span><span class="co">#&gt;  [11] pkgconfig_2.0.3               crayon_1.5.2                 </span></span>
<span><span class="co">#&gt;  [13] fastmap_1.1.1                 backports_1.4.1              </span></span>
<span><span class="co">#&gt;  [15] magick_2.8.3                  XVector_0.42.0               </span></span>
<span><span class="co">#&gt;  [17] ellipsis_0.3.2                labeling_0.4.3               </span></span>
<span><span class="co">#&gt;  [19] scuttle_1.12.0                utf8_1.2.4                   </span></span>
<span><span class="co">#&gt;  [21] promises_1.2.1                rmarkdown_2.26               </span></span>
<span><span class="co">#&gt;  [23] ragg_1.3.0                    purrr_1.0.2                  </span></span>
<span><span class="co">#&gt;  [25] bit_4.0.5                     xfun_0.42                    </span></span>
<span><span class="co">#&gt;  [27] beachmat_2.18.1               zlibbioc_1.48.2              </span></span>
<span><span class="co">#&gt;  [29] cachem_1.0.8                  jsonlite_1.8.8               </span></span>
<span><span class="co">#&gt;  [31] blob_1.2.4                    highr_0.10                   </span></span>
<span><span class="co">#&gt;  [33] later_1.3.2                   DelayedArray_0.28.0          </span></span>
<span><span class="co">#&gt;  [35] BiocParallel_1.36.0           interactiveDisplayBase_1.40.0</span></span>
<span><span class="co">#&gt;  [37] broom_1.0.5                   parallel_4.3.3               </span></span>
<span><span class="co">#&gt;  [39] R6_2.5.1                      bslib_0.6.1                  </span></span>
<span><span class="co">#&gt;  [41] car_3.1-2                     jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.12                   knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [45] httpuv_1.6.14                 Matrix_1.6-5                 </span></span>
<span><span class="co">#&gt;  [47] tidyselect_1.2.1              abind_1.4-5                  </span></span>
<span><span class="co">#&gt;  [49] yaml_2.3.8                    codetools_0.2-19             </span></span>
<span><span class="co">#&gt;  [51] curl_5.2.1                    lattice_0.22-5               </span></span>
<span><span class="co">#&gt;  [53] tibble_3.2.1                  withr_3.0.0                  </span></span>
<span><span class="co">#&gt;  [55] KEGGREST_1.42.0               shiny_1.8.0                  </span></span>
<span><span class="co">#&gt;  [57] evaluate_0.23                 desc_1.4.3                   </span></span>
<span><span class="co">#&gt;  [59] Biostrings_2.70.3             pillar_1.9.0                 </span></span>
<span><span class="co">#&gt;  [61] BiocManager_1.30.22           filelock_1.0.3               </span></span>
<span><span class="co">#&gt;  [63] carData_3.0-5                 generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [65] RCurl_1.98-1.14               BiocVersion_3.18.1           </span></span>
<span><span class="co">#&gt;  [67] sparseMatrixStats_1.14.0      munsell_0.5.0                </span></span>
<span><span class="co">#&gt;  [69] scales_1.3.0                  xtable_1.8-4                 </span></span>
<span><span class="co">#&gt;  [71] glue_1.7.0                    tools_4.3.3                  </span></span>
<span><span class="co">#&gt;  [73] BiocNeighbors_1.20.2          ggsignif_0.6.4               </span></span>
<span><span class="co">#&gt;  [75] fs_1.6.3                      cowplot_1.1.3                </span></span>
<span><span class="co">#&gt;  [77] grid_4.3.3                    tidyr_1.3.1                  </span></span>
<span><span class="co">#&gt;  [79] AnnotationDbi_1.64.1          colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [81] GenomeInfoDbData_1.2.11       cli_3.6.2                    </span></span>
<span><span class="co">#&gt;  [83] rappdirs_0.3.3                textshaping_0.3.7            </span></span>
<span><span class="co">#&gt;  [85] fansi_1.0.6                   viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt;  [87] S4Arrays_1.2.1                dplyr_1.1.4                  </span></span>
<span><span class="co">#&gt;  [89] gtable_0.3.4                  rstatix_0.7.2                </span></span>
<span><span class="co">#&gt;  [91] sass_0.4.9                    digest_0.6.35                </span></span>
<span><span class="co">#&gt;  [93] SparseArray_1.2.4             farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [95] rjson_0.2.21                  memoise_2.0.1                </span></span>
<span><span class="co">#&gt;  [97] htmltools_0.5.7               pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt;  [99] lifecycle_1.0.4               httr_1.4.7                   </span></span>
<span><span class="co">#&gt; [101] mime_0.12                     bit64_4.0.5                  </span></span>
<span><span class="co">#&gt; [103] MASS_7.3-60.0.1</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael Totty.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
