<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SpotSweeper">
<title>Getting started with 'SpotSweeper' â€¢ SpotSweeper</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with 'SpotSweeper'">
<meta property="og:description" content="SpotSweeper">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SpotSweeper</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting_started.html">Getting started with 'SpotSweeper'</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MicTott/SpotSweeper/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with 'SpotSweeper'</h1>
                        <h4 data-toc-skip class="author">Michael
Totty</h4>
            <address class="author_afil">
      Johns Hopkins Bloomberg School of Public Health, Baltimore, MD,
USA<br><h4 data-toc-skip class="author">Boyi
Guo</h4>
            <address class="author_afil">
      Johns Hopkins Bloomberg School of Public Health, Baltimore, MD,
USA<br><h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            <address class="author_afil">
      Johns Hopkins Bloomberg School of Public Health, Baltimore, MD,
USA<br><h4 data-toc-skip class="date">2024-02-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/MicTott/SpotSweeper/blob/HEAD/vignettes/getting_started.Rmd" class="external-link"><code>vignettes/getting_started.Rmd</code></a></small>
      <div class="d-none name"><code>getting_started.Rmd</code></div>
    </address>
</address>
</address>
</div>

    
    
<div class="section level3">
<h3 id="input-data-format">Input data format<a class="anchor" aria-label="anchor" href="#input-data-format"></a>
</h3>
<p>In the examples below, we assume the input data are provided as a <a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a>
(SPE) object. The outputs for spot-level outliers and artifacts are
stored in the <code>colData</code> of the SPE object.</p>
</div>
<div class="section level3">
<h3 id="spot-level-local-outlier-detection">Spot-level local outlier detection<a class="anchor" aria-label="anchor" href="#spot-level-local-outlier-detection"></a>
</h3>
<p>This is an example workflow showing how to detect and visualize local
outliers in 10X Genomics Visium data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MicTott/SpotSweeper" class="external-link">SpotSweeper</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># load  Maynard et al DLPFC daatset</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">STexampleData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/STexampleData/man/STexampleData.html" class="external-link">Visium_humanDLPFC</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># change from gene id to gene names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># show column data before SpotSweeper</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "barcode_id"   "sample_id"    "in_tissue"    "array_row"    "array_col"   </span></span>
<span><span class="co">#&gt; [6] "ground_truth" "cell_count"</span></span>
<span></span>
<span><span class="co"># drop out-of-tissue spots</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="va">spe</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">ground_truth</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>SpotSweeper can be run on an SPE object with the following code. This
outputs the <code>local_outliers</code> in the colData of the SPE
object. Selecting <code>data_output=TRUE</code> exports z-transformed QC
metrics as well.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Identifying the mitochondrial transcripts in our SpatialExperiment.</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Calculating QC metrics for each spot using scuttle</span></span>
<span><span class="va">spe</span><span class="op">&lt;-</span> <span class="fu">scuttle</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">spe</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode_id"            "sample_id"             "in_tissue"            </span></span>
<span><span class="co">#&gt;  [4] "array_row"             "array_col"             "ground_truth"         </span></span>
<span><span class="co">#&gt;  [7] "cell_count"            "sum"                   "detected"             </span></span>
<span><span class="co">#&gt; [10] "subsets_Mito_sum"      "subsets_Mito_detected" "subsets_Mito_percent" </span></span>
<span><span class="co">#&gt; [13] "total"</span></span>
<span></span>
<span><span class="co"># Identifying local outliers using SpotSweeper</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'sum'</span> ,<span class="st">'detected'</span>, <span class="st">"subsets_Mito_percent"</span><span class="op">)</span></span>
<span><span class="va">spe</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/localOutliers.html">localOutliers</a></span><span class="op">(</span><span class="va">spe</span>, </span>
<span>                    features<span class="op">=</span><span class="va">features</span>,</span>
<span>                    n_neighbors<span class="op">=</span><span class="fl">18</span>, </span>
<span>                    data_output<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                    method<span class="op">=</span><span class="st">"multivariate"</span></span>
<span>                    <span class="op">)</span></span>
<span></span>
<span><span class="co"># show column data after SpotSweeper</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode_id"                "sample_id"                </span></span>
<span><span class="co">#&gt;  [3] "in_tissue"                 "array_row"                </span></span>
<span><span class="co">#&gt;  [5] "array_col"                 "ground_truth"             </span></span>
<span><span class="co">#&gt;  [7] "cell_count"                "sum"                      </span></span>
<span><span class="co">#&gt;  [9] "detected"                  "subsets_Mito_sum"         </span></span>
<span><span class="co">#&gt; [11] "subsets_Mito_detected"     "subsets_Mito_percent"     </span></span>
<span><span class="co">#&gt; [13] "total"                     "sum_log2"                 </span></span>
<span><span class="co">#&gt; [15] "detected_log2"             "subsets_Mito_percent_log2"</span></span>
<span><span class="co">#&gt; [17] "coords"                    "local_outliers"           </span></span>
<span><span class="co">#&gt; [19] "sum_z"                     "detected_z"               </span></span>
<span><span class="co">#&gt; [21] "subsets_Mito_percent_z"    "LOF"</span></span></code></pre></div>
<p>We can now visualize <code>local_outliers</code> vs one of the QC
metrics, <code>sum_log2</code>, with help from the <code>escheR</code>
package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/boyiguo1/escheR" class="external-link">escheR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plotting using escheR</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/make_escheR.html" class="external-link">make_escheR</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_fill.html" class="external-link">add_fill</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"sum_log2"</span>, point_size<span class="op">=</span><span class="fl">1.25</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span><span class="st">"white"</span>,high <span class="op">=</span>  <span class="st">"darkgreen"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/make_escheR.html" class="external-link">make_escheR</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_fill.html" class="external-link">add_fill</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"sum_log2"</span>, point_size<span class="op">=</span><span class="fl">1.25</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/escheR/man/add_ground.html" class="external-link">add_ground</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"local_outliers"</span>, stroke <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">""</span>, <span class="co"># turn off legend name for ground_truth</span></span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>      <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"transparent"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient</a></span><span class="op">(</span>low <span class="op">=</span><span class="st">"white"</span>,high <span class="op">=</span>  <span class="st">"darkgreen"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>  plotlist <span class="op">=</span> <span class="va">plot_list</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">2</span>, nrow <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  common.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="getting_started_files/figure-html/local_outlier_plot-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggpubr_0.6.0                escheR_1.2.0               </span></span>
<span><span class="co">#&gt;  [3] ggplot2_3.4.4               STexampleData_1.10.0       </span></span>
<span><span class="co">#&gt;  [5] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.32.0 Biobase_2.62.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.54.1        GenomeInfoDb_1.38.6        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.36.0              S4Vectors_0.40.2           </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.14.0       matrixStats_1.2.0          </span></span>
<span><span class="co">#&gt; [15] ExperimentHub_2.10.0        AnnotationHub_3.10.0       </span></span>
<span><span class="co">#&gt; [17] BiocFileCache_2.10.1        dbplyr_2.4.0               </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.48.1         SpotSweeper_0.99.1         </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] DBI_1.2.1                     bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [3] rlang_1.1.3                   magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;   [5] compiler_4.3.2                RSQLite_2.3.5                </span></span>
<span><span class="co">#&gt;   [7] DelayedMatrixStats_1.24.0     png_0.1-8                    </span></span>
<span><span class="co">#&gt;   [9] systemfonts_1.0.5             vctrs_0.6.5                  </span></span>
<span><span class="co">#&gt;  [11] stringr_1.5.1                 pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [13] crayon_1.5.2                  fastmap_1.1.1                </span></span>
<span><span class="co">#&gt;  [15] backports_1.4.1               magick_2.8.2                 </span></span>
<span><span class="co">#&gt;  [17] XVector_0.42.0                ellipsis_0.3.2               </span></span>
<span><span class="co">#&gt;  [19] labeling_0.4.3                scuttle_1.12.0               </span></span>
<span><span class="co">#&gt;  [21] utf8_1.2.4                    promises_1.2.1               </span></span>
<span><span class="co">#&gt;  [23] rmarkdown_2.25                ragg_1.2.7                   </span></span>
<span><span class="co">#&gt;  [25] purrr_1.0.2                   bit_4.0.5                    </span></span>
<span><span class="co">#&gt;  [27] xfun_0.42                     beachmat_2.18.0              </span></span>
<span><span class="co">#&gt;  [29] zlibbioc_1.48.0               cachem_1.0.8                 </span></span>
<span><span class="co">#&gt;  [31] jsonlite_1.8.8                blob_1.2.4                   </span></span>
<span><span class="co">#&gt;  [33] highr_0.10                    later_1.3.2                  </span></span>
<span><span class="co">#&gt;  [35] DelayedArray_0.28.0           BiocParallel_1.36.0          </span></span>
<span><span class="co">#&gt;  [37] interactiveDisplayBase_1.40.0 broom_1.0.5                  </span></span>
<span><span class="co">#&gt;  [39] parallel_4.3.2                R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [41] bslib_0.6.1                   stringi_1.8.3                </span></span>
<span><span class="co">#&gt;  [43] car_3.1-2                     jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt;  [45] Rcpp_1.0.12                   knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [47] httpuv_1.6.14                 Matrix_1.6-1.1               </span></span>
<span><span class="co">#&gt;  [49] tidyselect_1.2.0              abind_1.4-5                  </span></span>
<span><span class="co">#&gt;  [51] yaml_2.3.8                    codetools_0.2-19             </span></span>
<span><span class="co">#&gt;  [53] curl_5.2.0                    lattice_0.21-9               </span></span>
<span><span class="co">#&gt;  [55] tibble_3.2.1                  shiny_1.8.0                  </span></span>
<span><span class="co">#&gt;  [57] withr_3.0.0                   KEGGREST_1.42.0              </span></span>
<span><span class="co">#&gt;  [59] evaluate_0.23                 desc_1.4.3                   </span></span>
<span><span class="co">#&gt;  [61] Biostrings_2.70.2             pillar_1.9.0                 </span></span>
<span><span class="co">#&gt;  [63] BiocManager_1.30.22           filelock_1.0.3               </span></span>
<span><span class="co">#&gt;  [65] carData_3.0-5                 generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [67] dbscan_1.1-12                 RCurl_1.98-1.14              </span></span>
<span><span class="co">#&gt;  [69] BiocVersion_3.18.1            munsell_0.5.0                </span></span>
<span><span class="co">#&gt;  [71] scales_1.3.0                  sparseMatrixStats_1.14.0     </span></span>
<span><span class="co">#&gt;  [73] xtable_1.8-4                  glue_1.7.0                   </span></span>
<span><span class="co">#&gt;  [75] tools_4.3.2                   BiocNeighbors_1.20.2         </span></span>
<span><span class="co">#&gt;  [77] ggsignif_0.6.4                fs_1.6.3                     </span></span>
<span><span class="co">#&gt;  [79] cowplot_1.1.3                 grid_4.3.2                   </span></span>
<span><span class="co">#&gt;  [81] tidyr_1.3.1                   colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [83] AnnotationDbi_1.64.1          GenomeInfoDbData_1.2.11      </span></span>
<span><span class="co">#&gt;  [85] cli_3.6.2                     rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [87] textshaping_0.3.7             fansi_1.0.6                  </span></span>
<span><span class="co">#&gt;  [89] viridisLite_0.4.2             S4Arrays_1.2.0               </span></span>
<span><span class="co">#&gt;  [91] dplyr_1.1.4                   gtable_0.3.4                 </span></span>
<span><span class="co">#&gt;  [93] rstatix_0.7.2                 sass_0.4.8                   </span></span>
<span><span class="co">#&gt;  [95] digest_0.6.34                 SparseArray_1.2.4            </span></span>
<span><span class="co">#&gt;  [97] farver_2.1.1                  rjson_0.2.21                 </span></span>
<span><span class="co">#&gt;  [99] memoise_2.0.1                 htmltools_0.5.7              </span></span>
<span><span class="co">#&gt; [101] pkgdown_2.0.7                 lifecycle_1.0.4              </span></span>
<span><span class="co">#&gt; [103] httr_1.4.7                    mime_0.12                    </span></span>
<span><span class="co">#&gt; [105] bit64_4.0.5                   MASS_7.3-60</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Michael Totty.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
